sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -139,7 +139,7 @@ object ScalaReflection extends ScalaReflection {
    * This method help us "remember" the required data type by adding a `UpCast`. Note that we
    * only need to do this for leaf nodes.
    */
-  private def upCastToExpectedType(expr: Expression, expected: DataType,
+  private[spark] def upCastToExpectedType(expr: Expression, expected: DataType,
```

Comment:
How about moving this function into `object Upcast {}`?

Reply:
Nice suggestion! Btw, it might be a part of refactor between `deserializerForType` so may not need to have its own object.